colexec: TestHashRouterRandom fails

I saw it on one of CI [builds](https://teamcity.cockroachdb.com/viewLog.html?buildId=1814084&buildTypeId=Cockroach_UnitTests) and it can be easily reproduced under stress.
```
Yahors-MacBook-Pro:cockroach yuzefovich$ make stress PKG=./pkg/sql/colexec TESTS=TestHashRouterRandom
Running make with -j16
GOPATH set to /Users/yuzefovich/go
GOFLAGS= go test   -exec 'stress ' -tags ' make x86_64_apple_darwin19.3.0' -ldflags '-X github.com/cockroachdb/cockroach/pkg/build.typ=development -extldflags "" -X "github.com/cockroachdb/cockroach/pkg/build.tag=v20.1.0-alpha.20191118-3237-g1139489888" -X "github.com/cockroachdb/cockroach/pkg/build.rev=1139489888fb15258efd7df01159f6ee3bd0bf49" -X "github.com/cockroachdb/cockroach/pkg/build.cgoTargetTriple=x86_64-apple-darwin19.3.0"  ' -run "TestHashRouterRandom" -timeout 0 ./pkg/sql/colexec  -v -args -test.timeout 20m

I200317 20:05:12.239051 1 rand.go:85  Random seed: -5198129252426211048
coldata.BatchSize() is set to 1024
=== RUN   TestHashRouterRandom
=== RUN   TestHashRouterRandom/numOutputs=18/blockedThreshold=831/outputSize=2962/totalInputSize=600/hashCols=[0_1]/cancel=false/memoryLimit=0_B
=== RUN   TestHashRouterRandom/numOutputs=18/blockedThreshold=831/outputSize=2962/totalInputSize=600/hashCols=[0_1]/cancel=false/memoryLimit=0_B/batchSize=1/sel=false
=== RUN   TestHashRouterRandom/numOutputs=18/blockedThreshold=831/outputSize=2962/totalInputSize=600/hashCols=[0_1]/cancel=false/memoryLimit=0_B/batchSize=1/sel=true
=== RUN   TestHashRouterRandom/numOutputs=18/blockedThreshold=831/outputSize=2962/totalInputSize=600/hashCols=[0_1]/cancel=false/memoryLimit=0_B/batchSize=2/sel=false
panic: test-disk: no bytes in account to release, current 581, free 661

goroutine 121 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Shrink(0xc0002f4560, 0x69f49e0, 0xc0000f8008, 0x295)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:568 +0x1a9
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Close(0xc00010d7a0, 0x69f49e0, 0xc0000f8008, 0xc0008e3e88, 0x582dcc2)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:428 +0x23d
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*spillingQueue).close(0xc0003a23c0, 0x69f49e0, 0xc0000f8008, 0x40312f1, 0x660d898)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/spilling_queue.go:263 +0x73
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*routerOutputOp).closeLocked(0xc000030ea0, 0x69f49e0, 0xc0000f8008)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:225 +0x4b
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*routerOutputOp).Next(0xc000030ea0, 0x69f49e0, 0xc0000f8008, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:218 +0x23b
github.com/cockroachdb/cockroach/pkg/sql/colexec.TestHashRouterRandom.func1.1.1(0xc00010cea0, 0x12, 0x12, 0x69f49e0, 0xc0000f8008, 0xc0002ae3f8, 0xc0001901b0, 0x12, 0x12, 0xc0002ae8e0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers_test.go:835 +0x79
created by github.com/cockroachdb/cockroach/pkg/sql/colexec.TestHashRouterRandom.func1.1
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers_test.go:833 +0x67c


ERROR: exit status 2

2 runs completed, 1 failures, over 3s
context canceled
```

I believe the problem is that although monitors can be accessed concurrently, accounts cannot - each router output needs to have its own disk account.

--- a/pkg/sql/colexec/routers.go
+++ b/pkg/sql/colexec/routers.go
@@ -386,7 +386,8 @@ type HashRouter struct {
 // by comparing memory use in the allocator with the memoryLimit argument. Each
 // Operator must have an independent allocator (this means that each allocator
 // should be linked to an independent mem account) as Operator.Next will usually
-// be called concurrently between different outputs.
+// be called concurrently between different outputs. Similarly, each output
+// needs to have a separate disk account.
 func NewHashRouter(
 	unlimitedAllocators []*Allocator,
 	input Operator,
@@ -395,7 +396,7 @@ func NewHashRouter(
 	memoryLimit int64,
 	diskQueueCfg colcontainer.DiskQueueCfg,
 	fdSemaphore semaphore.Semaphore,
-	diskAcc *mon.BoundAccount,
+	diskAccounts []*mon.BoundAccount,
 ) (*HashRouter, []Operator) {
 	if diskQueueCfg.CacheMode != colcontainer.DiskQueueCacheModeDefault {
 		execerror.VectorizedInternalPanic(errors.Errorf("hash router instantiated with incompatible disk queue cache mode: %d", diskQueueCfg.CacheMode))
@@ -412,7 +413,7 @@ func NewHashRouter(
 	unblockEventsChan := make(chan struct{}, 2*len(unlimitedAllocators))
 	memoryLimitPerOutput := memoryLimit / int64(len(unlimitedAllocators))
 	for i := range unlimitedAllocators {
-		op := newRouterOutputOp(unlimitedAllocators[i], types, unblockEventsChan, memoryLimitPerOutput, diskQueueCfg, fdSemaphore, diskAcc)
+		op := newRouterOutputOp(unlimitedAllocators[i], types, unblockEventsChan, memoryLimitPerOutput, diskQueueCfg, fdSemaphore, diskAccounts[i])
 		outputs[i] = op
 		outputsAsOps[i] = op
 	}

--- a/pkg/sql/colexec/routers_test.go
+++ b/pkg/sql/colexec/routers_test.go
@@ -27,6 +27,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/colcontainerutils"
 	"github.com/cockroachdb/cockroach/pkg/util/humanizeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/pkg/errors"
 	"github.com/stretchr/testify/require"
@@ -708,7 +709,13 @@ func TestHashRouterOneOutput(t *testing.T) {
 		t.Run(fmt.Sprintf("memoryLimit=%s", humanizeutil.IBytes(mtc.bytes)), func(t *testing.T) {
 			// Clear the testAllocator for use.
 			testAllocator.Clear()
-			r, routerOutputs := NewHashRouter([]*Allocator{testAllocator}, newOpFixedSelTestInput(sel, len(sel), data), typs, []uint32{0}, mtc.bytes, queueCfg, NewTestingSemaphore(2), testDiskAcc)
+			diskAcc := testDiskMonitor.MakeBoundAccount()
+			defer diskAcc.Close(ctx)
+			r, routerOutputs := NewHashRouter(
+				[]*Allocator{testAllocator}, newOpFixedSelTestInput(sel, len(sel), data),
+				typs, []uint32{0}, mtc.bytes, queueCfg, NewTestingSemaphore(2),
+				[]*mon.BoundAccount{&diskAcc},
+			)
 
 			if len(routerOutputs) != 1 {
 				t.Fatalf("expected 1 router output but got %d", len(routerOutputs))
@@ -809,12 +816,15 @@ func TestHashRouterRandom(t *testing.T) {
 				outputsAsOps := make([]Operator, numOutputs)
 				memoryLimitPerOutput := mtc.bytes / int64(len(outputs))
 				for i := range outputs {
+					// Create separate monitoring infrastructure as well as
+					// an allocator for each output as router outputs run
+					// concurrently.
 					acc := testMemMonitor.MakeBoundAccount()
 					defer acc.Close(ctx)
-					// Create a separate allocator for each output as a single allocator
-					// may not be used concurrently.
+					diskAcc := testDiskMonitor.MakeBoundAccount()
+					defer diskAcc.Close(ctx)
 					allocator := NewAllocator(ctx, &acc)
-					op := newRouterOutputOpWithBlockedThresholdAndBatchSize(allocator, typs, unblockEventsChan, memoryLimitPerOutput, queueCfg, NewTestingSemaphore(len(outputs)*2), blockedThreshold, outputSize, testDiskAcc)
+					op := newRouterOutputOpWithBlockedThresholdAndBatchSize(allocator, typs, unblockEventsChan, memoryLimitPerOutput, queueCfg, NewTestingSemaphore(len(outputs)*2), blockedThreshold, outputSize, &diskAcc)
 					outputs[i] = op
 					outputsAsOps[i] = op
 				}
@@ -912,12 +922,16 @@ func BenchmarkHashRouter(b *testing.B) {
 		for _, numInputBatches := range []int{2, 4, 8, 16} {
 			b.Run(fmt.Sprintf("numOutputs=%d/numInputBatches=%d", numOutputs, numInputBatches), func(b *testing.B) {
 				allocators := make([]*Allocator, numOutputs)
+				diskAccounts := make([]*mon.BoundAccount, numOutputs)
 				for i := range allocators {
 					acc := testMemMonitor.MakeBoundAccount()
 					allocators[i] = NewAllocator(ctx, &acc)
 					defer acc.Close(ctx)
+					diskAcc := testDiskMonitor.MakeBoundAccount()
+					diskAccounts[i] = &diskAcc
+					defer diskAcc.Close(ctx)
 				}
-				r, outputs := NewHashRouter(allocators, input, types, []uint32{0}, 64<<20, queueCfg, &TestingSemaphore{}, testDiskAcc)
+				r, outputs := NewHashRouter(allocators, input, types, []uint32{0}, 64<<20, queueCfg, &TestingSemaphore{}, diskAccounts)
 				b.SetBytes(8 * int64(coldata.BatchSize()) * int64(numInputBatches))
 				// We expect distribution to not change. This is a sanity check that
 				// we're resetting properly.

--- a/pkg/sql/colflow/vectorized_flow.go
+++ b/pkg/sql/colflow/vectorized_flow.go
@@ -95,10 +95,11 @@ type vectorizedFlow struct {
 	// the streaming components.
 	streamingMemAccounts []*mon.BoundAccount
 
-	// monitors are the memory monitors of the buffering components.
+	// monitors are the monitors (of both memory and disk usage) of the
+	// buffering components.
 	monitors []*mon.BytesMonitor
-	// acccounts are the memory accounts that are tracking the dynamic
-	// memory usage of the buffering components.
+	// accounts are the accounts that are tracking the dynamic memory and disk
+	// usage of the buffering components.
 	accounts []*mon.BoundAccount
 
 	tempStorage struct {
@@ -194,7 +195,7 @@ func (f *vectorizedFlow) Setup(
 		f.operatorConcurrency = creator.operatorConcurrency
 		f.streamingMemAccounts = append(f.streamingMemAccounts, creator.streamingMemAccounts...)
 		f.monitors = append(f.monitors, creator.monitors...)
-		f.accounts = append(f.accounts, creator.acccounts...)
+		f.accounts = append(f.accounts, creator.accounts...)
 		log.VEventf(ctx, 1, "vectorized flow setup succeeded")
 		return ctx, nil
 	}
@@ -204,7 +205,7 @@ func (f *vectorizedFlow) Setup(
 	for _, acc := range creator.streamingMemAccounts {
 		acc.Close(ctx)
 	}
-	for _, acc := range creator.acccounts {
+	for _, acc := range creator.accounts {
 		acc.Close(ctx)
 	}
 	for _, mon := range creator.monitors {
@@ -442,12 +443,12 @@ type vectorizedFlowCreator struct {
 	// streamingMemAccounts contains all memory accounts of the non-buffering
 	// components in the vectorized flow.
 	streamingMemAccounts []*mon.BoundAccount
-	// monitors contains all memory monitors of the buffering
-	// components in the vectorized flow.
+	// monitors contains all monitors (for both memory and disk usage) of the
+	// buffering components in the vectorized flow.
 	monitors []*mon.BytesMonitor
-	// acccounts contains all memory accounts of the buffering
-	// components in the vectorized flow.
-	acccounts []*mon.BoundAccount
+	// accounts contains all monitors (for both memory and disk usage) of the
+	// buffering components in the vectorized flow.
+	accounts []*mon.BoundAccount
 
 	diskQueueCfg colcontainer.DiskQueueCfg
 	fdSemaphore  semaphore.Semaphore
@@ -497,18 +498,22 @@ func (s *vectorizedFlowCreator) createBufferingUnlimitedMemMonitor(
 	return bufferingOpUnlimitedMemMonitor
 }
 
-// createDiskAccount instantiates an unlimited disk monitor and a disk account
+// createDiskAccounts instantiates an unlimited disk monitor and disk accounts
 // to be used for disk spilling infrastructure in vectorized engine.
-// TODO(azhng): consolidates all allocation monitors/account manage into one
+// TODO(azhng): consolidate all allocation monitors/account management into one
 // place after branch cut for 20.1.
-func (s *vectorizedFlowCreator) createDiskAccount(
-	ctx context.Context, flowCtx *execinfra.FlowCtx, name string,
-) *mon.BoundAccount {
+func (s *vectorizedFlowCreator) createDiskAccounts(
+	ctx context.Context, flowCtx *execinfra.FlowCtx, name string, numAccounts int,
+) []*mon.BoundAccount {
 	diskMonitor := execinfra.NewMonitor(ctx, flowCtx.Cfg.DiskMonitor, name)
 	s.monitors = append(s.monitors, diskMonitor)
-	diskAccount := diskMonitor.MakeBoundAccount()
-	s.acccounts = append(s.acccounts, &diskAccount)
-	return &diskAccount
+	diskAccounts := make([]*mon.BoundAccount, numAccounts)
+	for i := range diskAccounts {
+		diskAcc := diskMonitor.MakeBoundAccount()
+		diskAccounts[i] = &diskAcc
+	}
+	s.accounts = append(s.accounts, diskAccounts...)
+	return diskAccounts
 }
 
 // newStreamingMemAccount creates a new memory account bound to the monitor in
@@ -591,15 +596,16 @@ func (s *vectorizedFlowCreator) setupRouter(
 	for i := range allocators {
 		acc := hashRouterMemMonitor.MakeBoundAccount()
 		allocators[i] = colexec.NewAllocator(ctx, &acc)
-		s.acccounts = append(s.acccounts, &acc)
+		s.accounts = append(s.accounts, &acc)
 	}
 	limit := execinfra.GetWorkMemLimit(flowCtx.Cfg)
 	if flowCtx.Cfg.TestingKnobs.ForceDiskSpill {
 		limit = 1
 	}
-	diskAccount := s.createDiskAccount(ctx, flowCtx, mmName)
+	diskAccounts := s.createDiskAccounts(ctx, flowCtx, mmName, len(output.Streams))
 	router, outputs := colexec.NewHashRouter(
-		allocators, input, outputTyps, output.HashColumns, limit, s.diskQueueCfg, s.fdSemaphore, diskAccount)
+		allocators, input, outputTyps, output.HashColumns, limit, s.diskQueueCfg, s.fdSemaphore, diskAccounts,
+	)
 	runRouter := func(ctx context.Context, _ context.CancelFunc) {
 		logtags.AddTag(ctx, "hashRouterID", mmName)
 		router.Run(ctx)
@@ -923,7 +929,7 @@ func (s *vectorizedFlowCreator) setupFlow(
 		// monitor and account have been created, so we always want to accumulate
 		// them for a proper cleanup.
 		s.monitors = append(s.monitors, result.OpMonitors...)
-		s.acccounts = append(s.acccounts, result.OpAccounts...)
+		s.accounts = append(s.accounts, result.OpAccounts...)
 		if err != nil {
 			return nil, errors.Wrapf(err, "unable to vectorize execution plan")
 		}
@@ -1177,7 +1183,7 @@ func SupportsVectorized(
 		for _, acc := range creator.streamingMemAccounts {
 			acc.Close(ctx)
 		}
-		for _, acc := range creator.acccounts {
+		for _, acc := range creator.accounts {
 			acc.Close(ctx)
 		}
 		for _, mon := range creator.monitors {

--- a/pkg/sql/colflow/vectorized_flow_shutdown_test.go
+++ b/pkg/sql/colflow/vectorized_flow_shutdown_test.go
@@ -159,12 +159,19 @@ func TestVectorizedFlowShutdown(t *testing.T) {
 
 				// Create an allocator for each output.
 				allocators := make([]*colexec.Allocator, numHashRouterOutputs)
+				diskAccounts := make([]*mon.BoundAccount, numHashRouterOutputs)
 				for i := range allocators {
 					acc := testMemMonitor.MakeBoundAccount()
 					defer acc.Close(ctxRemote)
 					allocators[i] = colexec.NewAllocator(ctxRemote, &acc)
+					diskAcc := testDiskMonitor.MakeBoundAccount()
+					diskAccounts[i] = &diskAcc
+					defer diskAcc.Close(ctxRemote)
 				}
-				hashRouter, hashRouterOutputs := colexec.NewHashRouter(allocators, hashRouterInput, typs, []uint32{0}, 64<<20 /* 64 MiB */, queueCfg, &colexec.TestingSemaphore{}, testDiskAcc)
+				hashRouter, hashRouterOutputs := colexec.NewHashRouter(
+					allocators, hashRouterInput, typs, []uint32{0}, 64<<20, /* 64 MiB */
+					queueCfg, &colexec.TestingSemaphore{}, diskAccounts,
+				)
 				for i := 0; i < numInboxes; i++ {
 					inboxMemAccount := testMemMonitor.MakeBoundAccount()
 					defer inboxMemAccount.Close(ctxLocal)

